core/common/src/main/java/alluxio/conf/PropertyKey.java

```diff
@@ -2132,6 +2132,30 @@ public String toString() {
           .setConsistencyCheckLevel(ConsistencyCheckLevel.WARN)
           .setScope(Scope.MASTER)
           .build();
+  public static final PropertyKey MASTER_METADATA_SYNC_CONCURRENCY_LEVEL =
+      new Builder(Name.MASTER_METADATA_SYNC_CONCURRENCY_LEVEL)
+          .setDefaultValue(6)
+          .setDescription("The maximum number of concurrent sync tasks running for a given sync "
```

Comment:
How does this relate to the other parameters, MASTER_METADATA_SYNC_EXECUTOR_POOL_SIZE and MASTER_METADATA_SYNC_UFS_PREFETCH_POOL_SIZE? Also, it would be helpful to state what the potential tradeoffs are for a higher vs lower value (for the other parameters as well).

Reply:
Good question

- `MASTER_METADATA_SYNC_EXECUTOR_POOL_SIZE`
  - This parameter determines exactly how many inodes can be processing sync in parallel.
  - The executor affected by this value is shared by _all_ RPC calls to the master. So if you have 64 total threads in this executor, then a syncMetadata operation from a `listStatus` will compete for threads (submit tasks to the same executor) as a `createDirectory` RPC.

Now imagine you have a user who does a `listStatus` RPC that is high up in the inode tree, say, `/dir1`; but then another user comes along who wants to call `getStatus` on `/dir2`. It is mutually exclusive from `/dir1`, and may trigger a metadata sync, but if we allow `listStatus` on `/dir1` to use all threads within this executor, concurrent sync operations could suffer in performance, even if `/dir2` does not contain a large tree structure.

- `MASTER_METADATA_SYNC_CONCURRENCY_LEVEL` restricts max #of threads a single RPC is allowed to use at any given time for processing an inode sync.

The reason that `MASTER_METADATA_SYNC_UFS_PREFETCH_POOL_SIZE` is separate is because in order to make progress during sync, ufs prefetch threads must be available, and so if all threads in `MASTER_METADATA_SYNC_CONCURRENCY_LEVEL` are taken and none can be used for ufs fetching, then we could run into a deadlock scenario. There is no limit per sync operation on how many threads can be used for prefetch. Though the throughput of requests to the prefetch pool is limited to an extent by the `MASTER_METADATA_SYNC_CONCURRENCY_LEVEL`

Reply:
could we have some of these comments in the source code?  helps the next reader